wpilibc/src/main/native/cpp/simulation/DifferentialDrivetrainSim.cpp

```diff
@@ -0,0 +1,91 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "frc/simulation/DifferentialDrivetrainSim.h"
+
+#include "frc/system/RungeKutta.h"
+
+using namespace frc;
+using namespace frc::sim;
+
+DifferentialDrivetrainSim::DifferentialDrivetrainSim(
+    LinearSystem<2, 2, 2>& plant, DifferentialDriveKinematics& kinematics,
+    DCMotor leftGearbox, double leftGearing, DCMotor rightGearbox,
+    double rightGearing)
+    : m_plant(plant),
+      m_rb(kinematics.trackWidth / 2.0),
+      m_leftMotor(leftGearbox),
+      m_rightMotor(rightGearbox),
+      m_leftGearing(leftGearing),
+      m_rightGearing(rightGearing) {}
+
+void DifferentialDrivetrainSim::SetInputs(units::volt_t leftVoltage,
+                                          units::volt_t rightVoltage) {
+  m_u << leftVoltage.to<double>(), rightVoltage.to<double>();
+}
+
+void DifferentialDrivetrainSim::Update(units::second_t dt) {
+  m_x = RungeKutta([this](auto& x, auto& u) { return Dynamics(x, u); }, m_x,
+                   m_u, dt);
+}
+
+double DifferentialDrivetrainSim::GetState(int state) { return m_x(state); }
+
+Vector<10> DifferentialDrivetrainSim::GetState() { return m_x; }
+
+Rotation2d DifferentialDrivetrainSim::GetHeading() {
+  return Rotation2d(units::radian_t(m_x(State::kHeading)));
+}
+
+Pose2d DifferentialDrivetrainSim::GetEstimatedPosition() {
+  return Pose2d(units::meter_t(m_x(State::kX)), units::meter_t(m_x(State::kY)),
+                Rotation2d(units::radian_t(m_x(State::kHeading))));
+}
+
+units::ampere_t DifferentialDrivetrainSim::GetCurrentDrawAmps() {
+  auto loadIleft =
+      m_leftMotor.Current(
+          units::radians_per_second_t(m_x(State::kLeftVelocity) *
+                                      m_leftGearing / m_rb.to<double>()),
+          units::volt_t(m_u(0))) *
+      wpi::sgn(m_x(State::kLeftVelocity));
+
+  auto loadIRight =
+      m_rightMotor.Current(
+          units::radians_per_second_t(m_x(State::kRightVelocity) *
+                                      m_rightGearing / m_rb.to<double>()),
+          units::volt_t(m_u(1))) *
+      wpi::sgn(m_x(State::kRightVelocity));
+
+  return loadIleft + loadIRight;
```

Comment:
The current drawn from the battery is _not_ the same as the sum of the motor currents. Assuming the motor is not acting as a generator, this quantity should be always positive (current from the positive terminal to the negative terminal). The real-world equivalent would be to look at this as the current into the motor controller, not the current into the motor.

Since you're not modeling any of the behaviors of the motor controller this is likely good enough: `abs(ImotorLeft * dutyCycleLeft) + abs(ImotorRight * dutyCycleRight)`

Two examples:

1) The drivetrain is spinning in place, one motor has positive current, one motor has equivalent negative current. Right now this would report the drivetrain current as 0, but the battery would tell you otherwise.

2) This one might be less relevant. The robot has 0 voltage demand but non-0 velocity (momentum or being pushed). In this case the motor has non-0 current assuming brake mode, but the current seen by the battery should be 0.

There's a decent discussion on this here https://www.chiefdelphi.com/t/neo-motors-pulling-only-5-amps-when-stalled/350542

This needs to be done for the remaining simulation items as well.

Reply:
I think this was addressed in 41f3c1